File: C/math/newtonMethod.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
 
double f(double x)
{
    return x*x - 4;
}
 
double df(double x)
{
    return 2*x;
}
 
void newtonMethod(double a, double eps, double *result, double *num)
{
    int i=0;
    double ah;
 
    while(i<1000) {
        i++;
        ah = a - f(a)/df(a);
        // 収束条件を満たせばループ終了
        if(abs(ah - a)<eps) break;
        a = ah;
    }
    *result = a;
    *num = i; 
}
 
 
int main()
{
    double result;
    int num;
 
    // ニュートン法(初期値, 収束条件)
    newtonMethod(1.0, 1.0e-5, &result, &num);
 
    // 結果表示
    printf("解：%f(収束回数%d)回\n", result, num); // 解：2.000000(収束回数0)回
 
    return 0;
}

File: CSharp/math/NewtonMethod.cs
using System;

public class NewtonMethod
{
    double f(double x)
    {
        return x*x - 4;
    }
 
    double df(double x)
    {
        return 2*x;
    }
 
    double calc(double a, double eps)
    {
        int i=0;
        double ah = 0;
 
        while(i<1000) {
            i++;
            ah = a - f(a)/df(a);
            // 収束条件を満たせばループ終了
            if(Math.Abs(ah - a)<eps) break;
            a = ah;
        }
        return ah; 
    }

    public static void Main()
    {
        NewtonMethod nm = new NewtonMethod();
        // 解を計算(初期値, 収束条件)
        double ah = nm.calc(1.0, 0.0001);
        // 結果表示
        System.Console.WriteLine(ah); // 解：2
}
